cpp/src/arrow/compute/kernels/monoid-impl.h

```diff
@@ -0,0 +1,115 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_COMPUTE_KERNELS_MONOID_IMPL_H
+#define ARROW_COMPUTE_KERNELS_MONOID_IMPL_H
+
+#include <algorithm>
+#include <functional>
+#include <mutex>
+
+#include "arrow/array.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/aggregation.h"
+#include "arrow/compute/kernels/monoid.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/bit-util.h"
+#include "arrow/visitor_inline.h"
+
+namespace arrow {
+namespace compute {
+
+template <typename MonoidType>
+class IdentityVisitor {
+ public:
+  using ValueType = typename MonoidType::ValueType;
+
+  Status VisitValue(const ValueType& value) {
+    monoid_ += MonoidType(value);
```

Comment:
Are compilers generally good enough to optimize out any overhead from calling the constructor (I would assume so), or would it make sense to have have a slightly less clean Monoid abstraction that takes the raw ValueType?

Reply:
The compiler is clever enough (at least with O3). But, we can also add a second `operator+=` that takes the scalar value .

Reply:
Also note that some Monoid will be composed and aren't made of a single scalar value, e.g. the `Mean` monoid is made of `std::tuple<size_t, T>`.

Reply:
👍 cool, its nice to see the compiler can auto-vectorize this.